pkg/operator/sync_common.go

```diff
@@ -78,31 +82,79 @@ func manageControllerResources(c serviceCAOperator, resourcePath string, modifie
 	return nil
 }
 
-// TODO manage rotation in addition to initial creation
-func manageSignerCA(client coreclientv1.SecretsGetter, eventRecorder events.Recorder) (bool, error) {
+func manageSignerCA(client coreclientv1.SecretsGetter, eventRecorder events.Recorder, rawUnsupportedServiceCAConfig []byte) (bool, error) {
 	secret := resourceread.ReadSecretV1OrDie(v4_00_assets.MustAsset("v4.0.0/service-serving-cert-signer-controller/signing-secret.yaml"))
-	_, err := client.Secrets(secret.Namespace).Get(secret.Name, metav1.GetOptions{})
-	if !apierrors.IsNotFound(err) {
+
+	var existingCert *x509.Certificate
+	existing, err := client.Secrets(secret.Namespace).Get(secret.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// Secret will need to be created
+	} else if err != nil {
 		return false, err
+	} else {
+		// Secret exists - attempt to parse its certs
+		certData := existing.Data[corev1.TLSCertKey]
+		if len(certData) > 0 {
+			certs, err := cert.ParseCertsPEM(certData)
+			if err != nil {
+				return false, err
+			}
+			if len(certs) > 0 {
+				existingCert = certs[0]
+			}
+		}
 	}
+
+	rotationMsg := ""
+	if existingCert == nil {
+		// Secret does not exist or lacks the expected cert.
+		err = initializeSigningSecret(secret)
+		if err != nil {
+			return false, err
+		}
+	} else {
+		rotationMsg, err = rotateSigningSecret(existing, existingCert, rawUnsupportedServiceCAConfig)
```

Comment:
Why is it necessary to write the changes to `secret`? In the case of an existing secret, the expected form was previously only used if the secret did not exist or had a missing/malformed value for `tls.crt`.

`existing` is the result of a direct client call rather than being sourced from an informer, so there is no cache involved.